test/unit/tcache_max.c

```diff
@@ -360,10 +361,146 @@ TEST_BEGIN(test_thread_tcache_max) {
+	char inputs[100] = "8-128:1|160-160:11|170-8388609:0";
+	char *inputp = inputs;
+	expect_d_eq(mallctlbymib(mib_set, mib_set_len, NULL, NULL,
+	    (void *)&inputp, sizeof(char *)), 0,
+	    "Unexpected mallctlbymib() failure");
+	for (szind_t i = 0; i < TCACHE_NBINS_MAX; i++) {
+		if (i >= sz_size2index_compute(8) &&
+		    i <= sz_size2index_compute(128)) {
+			cache_bin_info_init(&tcache_bin_info[i], 1);
+		} else if (i == sz_size2index_compute(160)) {
+			cache_bin_info_init(&tcache_bin_info[i], 11);
+		} else if (i >= sz_size2index_compute(170)) {
+			cache_bin_info_init(&tcache_bin_info[i], 0);
+		}
+	}
+	check_bins_info(tcache_bin_info);
+
+	/*
+	 * Closes the tcache and set ncached_max of some bins.  It will be
+	 * set properly. Yet after Enabling the tcache, the new setting will
+	 * not be carried on.  Instead, the initial settings will be applied.
+	 */
+	bool e0 = false, e1;
+	size_t bool_sz = sizeof(bool);
+	expect_d_eq(mallctl("thread.tcache.enabled", (void *)&e1, &bool_sz,
+	    (void *)&e0, bool_sz), 0, "Unexpected mallctl() error");
+	expect_true(e1, "Unexpected previous tcache state");
+	strcpy(inputs, "0-112:8");
+	expect_d_eq(mallctlbymib(mib_set, mib_set_len, NULL, NULL,
+	    (void *)&inputp, sizeof(char *)), 0,
+	    "Unexpected mallctlbymib() failure");
+	/* All bin info is gone once tcache is disabled. */
+	for (szind_t i = 0; i < TCACHE_NBINS_MAX; i++) {
+		if (i >= sz_size2index_compute(0) &&
+		    i <= sz_size2index_compute(112)) {
+			cache_bin_info_init(&tcache_bin_info[i], 8);
+		} else {
+			cache_bin_info_init(&tcache_bin_info[i], 0);
+		}
+	}
+	check_bins_info(tcache_bin_info);
+
+	e0 = true;
+	expect_d_eq(mallctl("thread.tcache.enabled", (void *)&e1, &bool_sz,
+	    (void *)&e0, bool_sz), 0, "Unexpected mallctl() error");
+	expect_false(e1, "Unexpected previous tcache state");
+	memcpy(tcache_bin_info, tcache_bin_info_backup,
+	    sizeof(tcache_bin_info_backup));
+	check_bins_info(tcache_bin_info);
+
+	/* Settings should stay after resetting tcache_max. */
+	strcpy(inputs, "2048-6144:123");
+	expect_d_eq(mallctlbymib(mib_set, mib_set_len, NULL, NULL,
+	    (void *)&inputp, sizeof(char *)), 0,
+	    "Unexpected mallctlbymib() failure");
+	for (szind_t i = sz_size2index_compute(2048);
+	    i <= sz_size2index_compute(6144); i++) {
+		cache_bin_info_init(&tcache_bin_info[i], 123);
+	}
+	tsd_t *tsd = tsd_fetch();
+	tcache_t *tcache = tsd_tcachep_get(tsd);
+	assert(tcache != NULL);
+	tcache_slow_t *tcache_slow = tcache->tcache_slow;
+	size_t tcache_max = tcache_max_get(tcache_slow) + 10;
+	assert_d_eq(mallctl("thread.tcache.max",
+	    NULL, NULL, (void *)&tcache_max, sizeof(size_t)),.0,
+	    "Unexpected.mallctl().failure");
+	check_bins_info(tcache_bin_info);
+
+	return NULL;
+}
+
+TEST_BEGIN(test_ncached_max) {
+	test_skip_if(!config_stats);
+	test_skip_if(!opt_tcache);
+	test_skip_if(opt_prof);
+	test_skip_if(san_uaf_detection_enabled());
+	unsigned nthreads = 100;
```

Comment:
This is just for testing purposes, will make it smaller later.

Reply:
I usually add a comment like `/* TODO: change before merging. */` and later grep TODO in the final diff